### PR TITLE
Location full accuracy (issue #489)

### PR DESCRIPTION
**Pull request by adapptor-kurt**
Adds a LocationFullAccuracy pod for checking and requesting (temporary) full accuracy on iOS 14.  Addresses issue #489.

Updates the `request` rationale to require a `temporaryPurposeKey` for this permission.  I experimented with adding a separate function for this temporary permission, but I think this solution is more elegant.  Feedback welcome.

## Test Plan

Test on an iOS 14 beta 3 device in the example app with combinations of when-in-use/background and full/reduced accuracy permissions.

### What's required for testing (prerequisites)?

- iOS 14 beta 3
- Xcode Version 12.0 beta 3

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ❌     |

## Checklist

- [x] I have tested this on a device and a simulator
- [x] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [x] I updated the typed files (TS and Flow)
- [x] I added a sample use of the API in the example project (`example/App.js`)

### CODE REVIEW

**Comment by jaspermeijaard** [comment]
I think this will be a very good addition to the package 👍  🎉  One thing I found a bit inconvenient to work with is the way the required/optional rationele are implemented and break current implementations. Example of an implementation that now breaks:

```
async function askCameraPermissions() {
    const permission: "android.permission.CAMERA" | "ios.permission.CAMERA" = Platform.select({
        android: Permissions.PERMISSIONS.ANDROID.CAMERA,
        ios: Permissions.PERMISSIONS.IOS.CAMERA,
    })!;

    // Variable permission is not allowed as argument. Type error: "No overload matches this call."
    const result = await Permissions.request(permission); 

    return result === "granted";
}
```

Typescript output:
```
error TS2769: No overload matches this call.
  Overload 1 of 3, '(permission: Values<Readonly<{ ACCEPT_HANDOVER: "android.permission.ACCEPT_HANDOVER"; ACCESS_BACKGROUND_LOCATION: "android.permission.ACCESS_BACKGROUND_LOCATION"; ACCESS_COARSE_LOCATION: "android.permission.ACCESS_COARSE_LOCATION"; ... 25 more ...; WRITE_EXTERNAL_STORAGE: "android.permission.WRITE_EXTERNAL_STORAGE"; }>>, rationale?: Rationale | undefined): Promise<...>', gave the following error.
    Argument of type '"android.permission.CAMERA" | "ios.permission.CAMERA"' is not assignable to parameter of type 'Values<Readonly<{ ACCEPT_HANDOVER: "android.permission.ACCEPT_HANDOVER"; ACCESS_BACKGROUND_LOCATION: "android.permission.ACCESS_BACKGROUND_LOCATION"; ACCESS_COARSE_LOCATION: "android.permission.ACCESS_COARSE_LOCATION"; ... 25 more ...; WRITE_EXTERNAL_STORAGE: "android.permission.WRITE_EXTERNAL_STORAGE"; }>>'.
      Type '"ios.permission.CAMERA"' is not assignable to type 'Values<Readonly<{ ACCEPT_HANDOVER: "android.permission.ACCEPT_HANDOVER"; ACCESS_BACKGROUND_LOCATION: "android.permission.ACCESS_BACKGROUND_LOCATION"; ACCESS_COARSE_LOCATION: "android.permission.ACCESS_COARSE_LOCATION"; ... 25 more ...; WRITE_EXTERNAL_STORAGE: "android.permission.WRITE_EXTERNAL_STORAGE"; }>>'.
  Overload 2 of 3, '(permission: "ios.permission.CAMERA" | "ios.permission.APP_TRACKING_TRANSPARENCY" | "ios.permission.BLUETOOTH_PERIPHERAL" | "ios.permission.CALENDARS" | "ios.permission.CONTACTS" | ... 10 more ... | "ios.permission.STOREKIT"): Promise<...>', gave the following error.
    Argument of type '"android.permission.CAMERA" | "ios.permission.CAMERA"' is not assignable to parameter of type '"ios.permission.CAMERA" | "ios.permission.APP_TRACKING_TRANSPARENCY" | "ios.permission.BLUETOOTH_PERIPHERAL" | "ios.permission.CALENDARS" | "ios.permission.CONTACTS" | ... 10 more ... | "ios.permission.STOREKIT"'.
      Type '"android.permission.CAMERA"' is not assignable to type '"ios.permission.CAMERA" | "ios.permission.APP_TRACKING_TRANSPARENCY" | "ios.permission.BLUETOOTH_PERIPHERAL" | "ios.permission.CALENDARS" | "ios.permission.CONTACTS" | ... 10 more ... | "ios.permission.STOREKIT"'.

1298     const permissions = await Permissions.request(permission);
                                                       ~~~~~~~~~~
```

Any suggestion how to request permissions in an elegant way cross-platform?

**Comment by zoontek** [comment]
It's great, but we need to find a better API than this one.

**Comment by adapptor-kurt** [comment]
I restored the definition of `request` so that it doesn't break anything else.  But since there's no typing to enforce temporaryPurposeKey I added a runtime error when LOCATION_FULL_ACCURACY is requested without it.

@zoontek happy to change this around if you have some suggestions.  I also agree that it's not going to win any API design awards.

**Comment by zoontek** [comment]
Another solution would be making a breaking change (3.0.0) and enforcing usage of TypeScript 4 (with labelled tuples - https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/). It would also allow the renaming of checkNotifications / requestNotifications.

But this version juste came out, it's a bit extreme.

**Comment by adapptor-kurt** [comment]
@zoontek Are the new labelled tuples any different from function overloads that you can already do in TypeScript?  Either way, I think a separate function is better for check/requestNotifications because it's easier to document and understand for users of the package.  Function overloads are confusing, as demonstrated by me stuffing up the original version of this PR.

I had tried adding a new requestLocationTemporaryFullAccuracy function, but it seemed overkill for this iOS-only beta functionality.  The temporary purpose key also seemed to match the "rationale" parameter.  If Android had similar precision permissions, with the need for additional options, it might be worth it.

**Comment by zoontek** [comment]
@adapptor-kurt It can be used [like this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwK4QM4B4AqUIA9gQB2AJqlAPYBGAVhAMbAB8UAvFJgNoDWEI5AZuwC6AbgBQYuuUKpgUAJIB5AMqsoAbzFQoAQQAKegPqYASjoDCAaXkA5AOLGzN5Xp0mAojfMBNAFxQAcgBLclQAOkgAJwBbINRUEMIw-SNTC2t7Rx1nVw8vbwCAGi0oACEAGQBVd0xFRUwACUM9dxN5PQbWnXL-YNCIiBi4hOkwiura+qaWto6u8qKS827PABE3ZV6Q8KjY+MSw5fK1jcXtZYBZLq3+3eGDy66zqHNFG0wLTE3A7YGh-dGr3en2UzwAYhZ3IZ5KsbjtBnsRkkIeYoTDnuVFMtMEobIZugB1HTeb59eH-JFhTHY3H48pEkkYrE6HFvQxgyrlcr48zmSpmHxwv6Ig7Ulm0jlcnl8gUFYraMWsvEEzp42yGSrKdxCu4ApKK2kqzzQvGa9zPK6reQ6QzleSlMwmPw-W4I+6jS3W232x1ykoXeTmEyKDpvbUu8kij2B4Ohmzm+VQC71XE6t16sLJpXPUO1b0OtzOsnC91JXOKfO+54eAM2VatUm-XWUmu2esmUGJ5TyNppikHbttZ4udzucxNDyvOw2eRKvtRpIjscTseKaez3HD2oeayYeelsLKbfuXeLAC+UDgZCkMmA4kk0lkUBmAeU3beqjY6iUyjP99AkAKCoejpkiajwMgaDoABEACEByiMOIMHPqBiRqD+IH9tI-7gNAJgQAAjigshgkghAMIkOiRAA5mQLAlAAPlAHDNok+4ZgabKStyFgyhYcoUGAwCJKg-jqFAYBIJEYChBAli8P4siREEhDUVAZ5CIxzGsdI-juLgdDIMQEDoJhC6FBGJYccySrspyPG8vy-EBIwogSPwZEUdIUCRIRxHAAAFGEwVwDRolQPhRFoMApHkcJ0hUbRACUGhnhIvlRbIAUvvIb64soYQ-skBhZFYtgOGkORuJ4PhJSIUAAPQNVAiiWGIGX+dlrSvu+ziFSoVI2RK9nSk5PgWeJknSbJ8kgP4ABEBCyPN6l1Y1zWtelfnRV1Jg9fl-UFVmuITRJUkyagckKVAi3RStZ5rU1UA2PULVtR1O05XlH6HYNNJcSNvFjd4j3NS92CbUAA).

The confusion came from the mix-up with the Android rationale, which are not options but messages displayed to the user.

**Comment by adapptor-kurt** [comment]
@zoontek The overload version is [like this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwK4QM4B4AqUIA9gQB2AJqlAPYBGAVhAMbAB8UAvFJgNoDWEI5AZuwC6AbgBQYuuUKpgUAJIB5AMqsoAbzFQoAQQAKegPqYASjoDCAaXkA5AOLGzN5Xp0mAojfMBNAFxQAcgBLclQAOkgAJwBbINRUEMIw-SNTC2t7Rx1nVw8vbwCAGi0oACEAGQBVd0xFRUwACUM9dxN5PQbWnXL-YNCIiBi4hOkwiura+qaWto6u8qKS827PABE3ZV6Q8KjY+MSw5fK1jcXtZYBZLq3+3eGDy66zqHNFG0wLTE3A7YGh-dGr3en2UzwAYhZ3IZ5KsbjtBnsRkkIeYoTDnuVFMtMEobIZugB1HTeb59eH-JFhTHY3H48pEkkYrE6HFvQxgyrlcr48zmSpmHxwv6Ig7Ulm0jlcnl8gUFYraMWsvEEzp42yGSrKdxCu4ApKK2kqzzQvGa9zPK6reQ6QzleSlMwmPw-W4I+6jS3W232x1ykoXeTmEyKDpvbUu8kij2B4Ohmzm+VQC71XE6t16sLJpXPUO1b0OtzOsnC91JXOKfO+54eAM2VatUm-XWUmu2esmUGJ5TyNppikHbttZ4udzucxNDyvOw2eRKvtRpIjscTseKaez3HD2oeayYeelsLKbfuXeLAC+UDgZCkMmA4kk0lkUBmAeU3beqjY6iUyjP99AkAKCoejpkiajwMgaDoABEACEByiMOIMHPqBiRqD+IH9tI978EghAMGhkQQAAjigsgABTNok+4ZgabKStyFgyhYcoUGAwCJKg-jqFAYBIJEYChBAli8P4siREEhAAOZQGeACU-gAG7kEExDiLh+EcdIUBEaRaDAJRqHSP47i4HQyDEBA6CYQuhQRiWtHMkq7KcoxvL8ixASMApUDKap6l4QR2m6WRBlUcZKFYYQdnkOxnEAPzcbx-GCagwmiVA4mSTJ8kaGeEghfp5EvvIb64soYQ-skBhZFYtgOGkORuJ4PhySIUAAPQdVAiiWGIhUUSVZUfpVKhUk5EqudKHk+HZPF8QJQkiSA-gAEQELIq2yW1nXdb1BUkaFxWtK+77OKNFVZric3JYtaXLWtG3AFt8ntV1UA2PUPV9QNBlDWdFVVXReIMdNso7e9n3YPtQA).

I did have it like this, but made a mistake that broke @jaspermeijaard's case.

**Comment by adapptor-kurt** [comment]
temporaryPurposeKey is a reference to a message defined in the plist.  You can argue that this is also not an option, but a message displayed to the user.

**Comment by adapptor-kurt** [comment]
@zoontek I'd like to get this in, but if you have reservations about the API then I'm happy to make any suggested changes.  I think these are the choices:

1. Use `check`/`request` as is, with temporaryPurposeKey provided in the rationale argument.  The type is not enforced.
2. Use `check`/`request`, but with an overload specifically for LOCATION_FULL_ACCURACY.  I'll call the argument `options`.
3. Add a new check/request function for full-accuracy, like notifications.

File: ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    return resolve([RNPermissionHandlerLocationFullAccuracy getAccuracyStatus:locationManager]);
+  } else {
+    return resolve(RNPermissionStatusAuthorized);
+  }
+}
+
+- (void)requestWithResolver:(void (^ _Nonnull)(RNPermissionStatus))resolve
+                   rejecter:(void (^ _Nonnull)(NSError * _Nonnull))reject
+                  rationale:(NSDictionary *_Nullable)rationale {
+  if (!CLLocationManager.locationServicesEnabled || CLLocationManager.authorizationStatus != RNPermissionStatusAuthorized) {
```
Comment (zoontek) [inline ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m]: You can't compare `CLLocationManager.authorizationStatus` (Apple API) with `RNPermissionStatusAuthorized` (RNPermissions API)

File: ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                                        code:RNPermissionHandlerLocationFullAccuracyNoPurposeKey
+                                    userInfo:@{NSLocalizedDescriptionKey: @"temporaryPurposeKey is required to request LOCATION_FULL_ACCURACY"}]);
+    }
+
+    [locationManager requestTemporaryFullAccuracyAuthorizationWithPurposeKey:purposeKey
+                                                                  completion:^(NSError * _Nullable error) {
+      RNPermissionStatus status = [RNPermissionHandlerLocationFullAccuracy getAccuracyStatus:locationManager];
+
+      // Ignore errors due to full accuracy already being authorized
+      if (error && (error.code != kCLErrorPromptDeclined || status != RNPermissionStatusAuthorized)) {
```
Comment (zoontek) [inline ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m:71]: Isn't it `(error && error.code != kCLErrorPromptDeclined) || status != RNPermissionStatusAuthorized` ?

**Comment by zoontek** [comment]
> @zoontek I'd like to get this in, but if you have reservations about the API then I'm happy to make any suggested changes. I think these are the choices:
> 
> 1. Use `check`/`request` as is, with temporaryPurposeKey provided in the rationale argument.  The type is not enforced.
> 2. Use `check`/`request`, but with an overload specifically for LOCATION_FULL_ACCURACY.  I'll call the argument `options`.
> 3. Add a new check/request function for full-accuracy, like notifications.

A fourth option is possible: enforcing `full-accuracy` as temporary purpose key for everyone. It's slightly more restrictive (one message for full accuracy permission request per app), but it introduces no changes on the JS API.

Later (when iOS 14 will be widely available), it will be nice to rethink the API in a major `react-native-permissions` which handle correctly all the new permissions flow for the platform (https://github.com/react-native-community/react-native-permissions/issues/508).

**Comment by jaspermeijaard** [comment]
Keep up the good work. I think you guys are best aware of the requirements but defining a rest type based on the permission might allow you to make the second argument required/optional? [See this example](https://www.typescriptlang.org/play?ssl=11&ssc=25&pln=11&pc=38#code/C4TwDgpgBAChBOBbAlgZ1cg9gOygXigCgoSoAfKAIi1QDpIk0MdaYAJAeQBUOB9AGQCSAIQBKAQVEBNSsVIVqmOgxTos2Wvw4BhcV0EcAcrwBiAVX79e47drMTtMwoVCQoogIbB1EADbQCAG8oMABXeDAlCABpCBAALihUYHhkbABzKABfIhdwaDhGNRwOMG8cVAAeLigIAA9gCGwAE1RYBFVmbAA+fCga+saWtsVlDqZ1TR09A2NzS2tbexsZKAB+KABtT3LsPwgAXShE7a8ffzWD5wBjCuAoeAgAR1CIZL7q2oam1vairu6AAoVBMcIkuAAaKC0GGYMrqVCJQqddSlXZVLjdACU+F6gSyzkeLzewEBo3o42KGnY3D4QjEkhkWMIRNeyTJNAp-0mNJ4AhEEmklChwTCESisQSVA8ACNrpRsszWSSOUouSiWFpdPojKYLFYbHYHDIRSFwpFUDE4olKLL5YqWc82aTySCqVNtbM9QtDctHJQsUA).

**Comment by adapptor-kurt** [comment]
> A fourth option is possible: enforcing `full-accuracy` as temporary purpose key for everyone. It's slightly more restrictive (one message for full accuracy permission request per app), but it introduces no changes on the JS API.

I don't mind adding a default purpose key, but I already have a use case in an app I'm working on for more than one message.

> I think you guys are best aware of the requirements but defining a rest type based on the permission might allow you to make the second argument required/optional? [See this example](https://www.typescriptlang.org/play?ssl=11&ssc=25&pln=11&pc=38#code/C4TwDgpgBAChBOBbAlgZ1cg9gOygXigCgoSoAfKAIi1QDpIk0MdaYAJAeQBUOB9AGQCSAIQBKAQVEBNSsVIVqmOgxTos2Wvw4BhcV0EcAcrwBiAVX79e47drMTtMwoVCQoogIbB1EADbQCAG8oMABXeDAlCABpCBAALihUYHhkbABzKABfIhdwaDhGNRwOMG8cVAAeLigIAA9gCGwAE1RYBFVmbAA+fCga+saWtsVlDqZ1TR09A2NzS2tbexsZKAB+KABtT3LsPwgAXShE7a8ffzWD5wBjCuAoeAgAR1CIZL7q2oam1vairu6AAoVBMcIkuAAaKC0GGYMrqVCJQqddSlXZVLjdACU+F6gSyzkeLzewEBo3o42KGnY3D4QjEkhkWMIRNeyTJNAp-0mNJ4AhEEmklChwTCESisQSVA8ACNrpRsszWSSOUouSiWFpdPojKYLFYbHYHDIRSFwpFUDE4olKLL5YqWc82aTySCqVNtbM9QtDctHJQsUA).

Ah, I forgot that conditional types are distributive, so this does the right thing 👍

File: ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m (hunk 1 above)
Comment (adapptor-kurt) [reply ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m]: 🤦    good spotting, not sure what I was thinking.

File: ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m (hunk 2 above)
Comment (adapptor-kurt) [reply ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m:71]: No, a failure to get authorisation should still resolve the promise successfully.

What's happening here is that requestTemporaryFullAccuracyAuthorizationWithPurposeKey will error with kCLErrorPromptDeclined when permission is already granted, causing it not to show a prompt to the user.  When we see this error, and can confirm that authorisation is granted, the call is treated as a success instead.

**Comment by adapptor-kurt** [comment]
I have some changes which I think provide a reasonable level of type safety without making things too complicated.  I'll use "full-accuracy" as the default purpose key so that the options are no longer required, and then change the definition to this:

```typescript
function request<T extends Permission = Permission>(
  permission: T,
  options?: T extends typeof PERMISSIONS.IOS.LOCATION_FULL_ACCURACY
    ? FullAccuracyOptionsIOS
    : Rationale,
): Promise<PermissionStatus>;
```

How's this look?

**Comment by adapptor-kurt** [comment]
@zoontek the code changes are up, just waiting for some feedback before updating documentation.

**Comment by jaspermeijaard** [comment]
This API looks good and doesn't seem to break current permission requests. I'll continue testing when it's more final.

**Comment by 27leaves** [comment]
Do you have a suggestion on how to reference it in an existing project? I tried to modify the `package.js`like `"react-native-permissions": "react-native-community/react-native-permissions#503/head"`, but it does not work like this because the lib files are not included. I really need it in an existing project this week and it would be awesome to try and give feedback 👍

**Comment by jaspermeijaard** [comment]
> Do you have a suggestion on how to reference it in an existing project? I tried to modify the `package.js`like `"react-native-permissions": "react-native-community/react-native-permissions#503/head"`, but it does not work like this because the lib files are not included. I really need it in an existing project this week and it would be awesome to try and give feedback 👍

Try using the branch directly from adapptor-kurt his fork, [see NPM docs](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies).

**Comment by toshe** [comment]
@creat-or you can try this as suggested by @jaspermeijaard 

`"react-native-permissions": "https://github.com/adapptor-kurt/react-native-permissions#568f153"`

**Comment by jaspermeijaard** [comment]
@adapptor-kurt When I check this new permission in an iOS 14.0 simulator of Xcode version 12.0 beta 6, regardless of the value of  "Precise Location" in the Settings the check always resolves to `"unavailable"`. I should be able to check the permission first before requesting it right? Example:

```
await Permissions.check(Permissions.PERMISSIONS.IOS.LOCATION_FULL_ACCURACY) // => "unavailable" 
```

**Comment by 27leaves** [comment]
Thank you @jaspermeijaard and @toshe!

@jaspermeijaard Yes, I have the same behaviour and also wondering if we miss something.

**Comment by jaspermeijaard** [comment]
I think status `unavailable` makes sense when you don't enable the pod 😅   

@creat-or You need to include the pod in your project, please review your `Podfile` and `Info.plist` [see documentation](https://github.com/react-native-community/react-native-permissions#ios).

**Comment by 27leaves** [comment]
@jaspermeijaard 🙈 ... awesome, that worked, thank you! :)

**Comment by ezeferex** [comment]
@jaspermeijaard @creat-or Hi! I pulled down from that branch into my project and included the pod, but when I tried to compile shows me the next error. Are you getting the same one?
```
❌  /Users/ezequiel/Projects/ReactNative/mobile/node_modules/react-native-permissions/ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m:28:27: property 'accuracyAuthorization' not found on object of type 'CLLocationManager *'
  switch (locationManager.accuracyAuthorization) {
                          ^
❌  /Users/ezequiel/Projects/ReactNative/mobile/node_modules/react-native-permissions/ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m:71:35: use of undeclared identifier 'kCLErrorPromptDeclined'
      if (error && (error.code != kCLErrorPromptDeclined || status != RNPermissionStatusAuthorized)) {
                                  ^
❌  /Users/ezequiel/Projects/ReactNative/mobile/node_modules/react-native-permissions/ios/LocationFullAccuracy/RNPermissionHandlerLocationFullAccuracy.m:66:22: no visible @interface for 'CLLocationManager' declares the selector 'requestTemporaryFullAccuracyAuthorizationWithPurposeKey:completion:'
    [locationManager requestTemporaryFullAccuracyAuthorizationWithPurposeKey:purposeKey
     ~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by adapptor-kurt** [comment]
@ezeferex the LocationFullAccuracy pod requires Xcode 12.  The GM seed is available from Apple's developer portal.

For everyone that's tried it, has it all been working ok?  The documentation isn't particularly helpful so I could improve that, but most of these permissions already require you to read a lot of external documentation to understand how to use them.

**Comment by jaspermeijaard** [comment]
> @ezeferex the LocationFullAccuracy pod requires Xcode 12. The GM seed is available from Apple's developer portal.
> 
> For everyone that's tried it, has it all been working ok? The documentation isn't particularly helpful so I could improve that, but most of these permissions already require you to read a lot of external documentation to understand how to use them.

@adapptor-kurt Tested on several OS versions with several people and everything seem to work good. Thank you! As far as I am concerned, green light for merging 👍

**Comment by tanishkbajaj** [comment]
By when do you think the code will get merged and permissions library would be ready for use with IOS 14 changes?

**Comment by zoontek** [comment]
@tanishkbajaj Frankly, I don't know. I planned to work on it all day this Saturday, put cannot promise it will be done on time. A lot of things has to be treated: Checking for support from iOS 10 to iOS 14, add support for Android 11 too (and check for support from Android 4.1 to Android 11!), ensure that we don't miss new features that could introduces another required breaking changes later, check that the APIs for all these new features / permissions play well together, write a migration guide, etc.

I currently work alone on this, do it on my spare time and for free: even if a lot of companies seems to use this package, I got something like 20$ a month (from a personal friend and my company 😅).

Understand me: I don't do OSS for money or glory and I'm not a company with a clear roadmap trying to sell you some service too. That's what open source is 🙂

Sure, I will do my best releasing a beta for this new version as soon as possible. But meanwhile, you can still checkout @adapptor-kurt branch, use https://github.com/ds300/patch-package or even copy and paste new permission handlers directly in your project.

**Comment by mikehardy** [comment]
@tanishkbajaj behold the beauty of patch-package! https://github.com/ds300/patch-package

Integrate the PR, and enjoy

**Comment by tanishkbajaj** [comment]
> @tanishkbajaj Frankly, I don't know. I planned to work on it all day this Saturday, put cannot promise it will be done on time. A lot of things has to be treated: Checking for support from iOS 10 to iOS 14, add support for Android 11 too (and check for support from Android 4.1 to Android 11!), ensure that we don't miss new features that could introduces another required breaking changes later, check that the APIs for all these new features / permissions play well together, write a migration guide, etc.
> 
> I currently work alone on this, do it on my spare time and for free: even if a lot of companies seems to use this package, I got something like 20$ a month (from a personal friend and my company 😅).
> 
> Understand me: I don't do OSS for money or glory and I'm not a company with a clear roadmap trying to sell you some service too. That's what open source is 🙂
> 
> Sure, I will do my best releasing a beta for this new version as soon as possible. But meanwhile, you can still checkout @adapptor-kurt branch, use https://github.com/ds300/patch-package or even copy and paste new permission handlers directly in your project.


@zoontek I understand your point. Hats off to the hard work you guys put, hope I can contribute too next time. Also meanwhile I can checkout and look at @adapptor-kurt branch.

**Comment by 27leaves** [comment]
I have a problem now with another permission while using react-native-camera. When I use this PR with `https://github.com/adapptor-kurt/react-native-permissions#13bceb138505454890e3f88e516622c6736d3d93`, I get the following error when trying to get CAMERA access:

> RNPermissions.request was called with 1 arguments but expects 2 arguments. If you haven't changed this method yourself, this usually means that your versions of the native code and JavaScript code are out of sync. Updating both should make this error go away.

With the latest stable version the camera works as expected.
Am I missing something?

File: src/module.ios.ts [hunk 3]
```diff
@@ -31,9 +35,12 @@ async function check(permission: Permission): Promise<PermissionStatus> {
     : RESULTS.UNAVAILABLE;
 }
 
-async function request(permission: Permission): Promise<PermissionStatus> {
+async function request(
+  permission: Permission,
+  options?: RequestOptions,
```
Comment (zoontek) [inline src/module.ios.ts:40]: You have to default to empty object in case of undefined `options`
`options: RequestOptions = {},`

**Comment by adapptor-kurt** [comment]
I can't reproduce this error in the react-native-permissions example app.  When options is `undefined` this is passed as a `nil` argument to native code.

@creat-or Could you provide some more details about what you're testing with (RN, OS, etc versions).  Have you tried a clean build, maybe by also reinstalling your node_modules and pods?

**Comment by 27leaves** [comment]
> I can't reproduce this error in the react-native-permissions example app. When options is `undefined` this is passed as a `nil` argument to native code.
> 
> @creat-or Could you provide some more details about what you're testing with (RN, OS, etc versions). Have you tried a clean build, maybe by also reinstalling your node_modules and pods?

I managed to reproduce the behavior in a minimal example with the plugins that I use:

https://github.com/creat-or/MinimalQRCodeScanner

The versions, testing on a real iPhone7 with iOS14 (mac OS 10.15.6, XCode 12)
```
"react": "16.13.1",
    "react-native": "0.63.2",
    "react-native-camera": "^3.40.0",
    "react-native-permissions": "https://github.com/adapptor-kurt/react-native-permissions#13bceb138505454890e3f88e516622c6736d3d93",
    "react-native-qrcode-scanner": "moaazsidat/react-native-qrcode-scanner#v1.5.0"
```

**Comment by adapptor-kurt** [comment]
@creat-or your crash is caused by having two copies of react-native-permissions.  Yarn is failing to realise that my fork is compatible with the one used by react-native-qrcode-scanner:

```
$ find . -name react-native-permissions
./node_modules/react-native-permissions
./node_modules/react-native-qrcode-scanner/node_modules/react-native-permissions
```

When react-native-qrcode-scanner tries to call the API, it's calling the old API with a single parameter but it ends up invoking the pod from the fork that expects two.  You can fix this by forcing the resolution in package.json:

```
"resolutions": {
  "react-native-qrcode-scanner/react-native-permissions": "https://github.com/adapptor-kurt/react-native-permissions#13bceb138505454890e3f88e516622c6736d3d93"
},
```

If this PR is merged and released with a compatible minor version, then yarn should resolve this properly and the forced resolution shouldn't be necessary.

File: src/module.ios.ts (hunk 3 above)
Comment (adapptor-kurt) [reply src/module.ios.ts:40]: See below for the explanation of the crash

**Comment by 27leaves** [comment]
Thank you very much for your great assistance! Now the camera permissions work again 👍 Thanks!!

**Comment by lorenc-tomasz** [comment]
Hi,
any timeframe for this PR?

**Comment by flochtililoch** [comment]
> Hi,
> any timeframe for this PR?

In case you haven't, see this comment: https://github.com/zoontek/react-native-permissions/pull/503#issuecomment-698198032

**Comment by zoontek** [comment]
@adapptor-kurt Sorry for the delay and thanks for the good work.
I'm merging this in the upcoming 3.0.0 branch to continue working on it.

**Comment by adapptor-kurt** [comment]
Thanks @zoontek!  Apple and Google don't make it easy, so your hard work putting everything together and testing is really appreciated.

@creat-or since this is going into a v3.0.0, that indicates a breaking change with v2.  I would recommend continue using my branch until v3.0.0 is released and `react-native-qrcode-scanner` is also updated to use it.

**Comment by cdavie-weconnect** [comment]
Thanks for this update! @zoontek: Do you have a rough guess of how long before the 3.0.0 branch might be released? Like, do you expect it to be days, weeks, or months? My team is trying to decide whether to just wait for the official release vs using a branch or fork sooner. Thanks again!

**Comment by zoontek** [comment]
For those who missed it, I published a first beta of the next major release. Feedbacks are welcomed 🙂

```sh
yarn add react-native-permissions@next
```

A migration guide is available here: https://github.com/zoontek/react-native-permissions/blob/3.0.0/MIGRATION.md

**Pull request merged**